### ISSUE TITLE
Allow for the use of logweights instead of weights

### ISSUE THREAD

**Issue by AdrienCorenflos** [PythonOT/POT#173]
While this may not have much impact for numpy based calculations, any autodiff implementation would benefit greatly (in terms of numerical stability) from allowing users to feed log_weights directly instead of weights. This is particularly valid for a few reasons:

- Most packages out there prefer outputting logpdf for random variables
- Most techniques (bregman projections, smooth ot at least) can (should?) be written purely in the logspace without much work for numerical speed and stability

**Comment by AdrienCorenflos** [comment]
FYI this is what I mean for the log treatment. It doesn't work end to end (I didn't spend much time on this yet)
https://github.com/AdrienCorenflos/POT

**Comment by rflamary** [comment]
OK now i understand better the original comment.

Yes implementing Sinkhorn in log space is a good idea and we do not have it yet in POT (our stabilized still use dot).
This should be another implementation of sinkhorn instead of a replacement as you implemented (you could  call it 'log_sinkhorn' for instance).

The reason i want to keep both is that i'm not really convinced that the performance hit for log space is so small. dot is implemented in paralel in blas but I doubt logsumexp is so you can take an important hit on numpy implementation. Could you do a small sctript and compare the computation times for both?

**Comment by AdrienCorenflos** [comment]
Yes I still need to iron it out and compare. Bit busy at the moment to be
honest but I'll finish this when I can.

Fundamentally there will be a trade-off between logsumexp vs dot and one
between summing and dot. Not sure which one will win.

Regarding the BLAS point, I've been trying to see if there was something
implementing einsums in log-exp format but I couldn't find it.
On the other hand, summing is parallelized in BLAS and exponentiation is
vectorised in numpy, so it might just do the trick.
